jlm/llvm/opt/alias-analyses/OnlineCycleDetection.hpp

```diff
@@ -0,0 +1,368 @@
+      if (i + 1 < topologicalOrder.size() && sccIndex[node] == sccIndex[topologicalOrder[i + 1]])
+      {
+        // We know that the SCC consists of only roots, since GetSuccessors_ returns only roots
+        JLM_ASSERT(Set_.IsUnificationRoot(topologicalOrder[i + 1]));
+        // Make the next object the root, to keep the unification going
+        topologicalOrder[i + 1] = UnifyPointerObjects_(node, topologicalOrder[i + 1]);
+      }
+      else
+      {
+        // Add this root to the next index in the topological order
+        ObjectToTopoOrder_[node] = TopoOrderToObject_.size();
+        TopoOrderToObject_.push_back(node);
+      }
+    }
+
+#ifdef JLM_ENABLE_ASSERTS
+    ValidateTopologicalOrder();
+#endif
+  }
+
+  /** Call this function after adding a superset edge subset -> superset.
+   * Both must be unification roots.
+   * If subset has a higher topological index than superset, the topological order is re-arranged.
+   * If a cycle is detected, then it will be unified away, and the unification root is returned.
+   * @param subset the tail of the just added superset edge
+   * @param superset the head of the just added superset edge
+   * @return the root of the new unification, if any, otherwise nullopt.
+   */
+  std::optional<PointerObjectIndex>
+  MaintainTopologicalOrder(PointerObjectIndex subset, PointerObjectIndex superset)
+  {
+    JLM_ASSERT(Set_.IsUnificationRoot(subset) && Set_.IsUnificationRoot(superset));
+    // All unification roots should have positions in the topological order
+    JLM_ASSERT(ObjectToTopoOrder_[subset] != -1 && ObjectToTopoOrder_[superset] != -1);
+
+    // If adding the simple edge subset -> superset does not break the invariant, return
+    if (ObjectToTopoOrder_[subset] <= ObjectToTopoOrder_[superset])
+      return std::nullopt;
+
+    // Otherwise, we need to do reordering in this range to fix the invariant
+    const auto lowerTopo = ObjectToTopoOrder_[superset];
+    const auto upperTopo = ObjectToTopoOrder_[subset];
+
+    // Perform DFS to find all roots reachable from superset
+    // We only care about nodes where with its topo index <= upperTopo
+    // These nodes will all need to be moved after subset in the new topological order
+    JLM_ASSERT(DfsStack_.empty());
+    // All nodes that have been pushed to the dfs stack
+    DfsNodesVisited_.Clear();
+    // All nodes that could reach subset and thus are part of a subset -> superset -> subset cycle
+    DfsCycleNodes_.Clear();
+
+    // Start DFS from the superset
+    DfsStack_.emplace(superset, false);
+
+    while (!DfsStack_.empty())
+    {
+      const auto node = DfsStack_.top().first;
+      const bool firstVisit = !DfsStack_.top().second;
+      DfsStack_.pop();
+      JLM_ASSERT(Set_.IsUnificationRoot(node));
+
+      if (firstVisit)
+      {
+        // If this node has already been visited, skip it
+        if (!DfsNodesVisited_.Insert(node))
+          continue;
```

Comment:
Are the comments misleading or is something off here? How can it be that this is the firstVisit and then we check whether the node has already been visited?